[sil-combine] Fix for Ownership round 2

gottesmm opened the pull request:
Just splitting this up into multiple PRs to ease review. Going to be uploading a bunch of patches here.

gottesmm committed:
[sil-combine] Update pointer_to_address canonicalizations for OSSA.

For now I am not performing the [strict] elimination until I bring up the
interior pointer OSSA utilities. I want to make further progress before I loop
back around to that.

gottesmm committed:
[ownership] Make ForwardingOperand truly a loose wrapper around an operand.

This implies making -> and * return an Operand * instead of an
OwnershipForwardingInst. So one can thus do:

ForwardingOperand op;
op.myForwardingOperandMethod();
op->myOperandMethod();

gottesmm committed:
[ownership] Add some helpers to BorrowedValue that lets you use it easily as a value.

Specifically I provided a '->' and a '*' operator. So now you can do:

```
borrowedValue->callSILValueMethod();
```

and explicitly cast to SILValue using '*'.

gottesmm committed:
[ownership] Add a helper API on InteriorPointerOperand that lets you infer the InteriorPointerOperand from the interior pointer operand's user instruction's result.

Useful when needing to walk use->def and get this info.

gottesmm committed:
[ownership] Add helper methods to InteriorPointerOperand that let you easily use its internal operand.

Specifically, I added a '*' and a '->' operator so one can cast explicitly using
'*' and access Operand methods using '->'.

gottesmm committed:
[ownership] Add a new helper InteriorPointerOperand::getSingleBaseValue().

This returns the underlying BorrowedValue that introduce's the interior pointer
operand's associated value.

gottesmm committed:
[sil-combine] Add an OwnershipFixupContext to SILCombine.

gottesmm committed:
[sil] Define ValueBase::getNextInstruction().

Sometimes when you are working with SILValues, you need a "next" insertion
point. This creates a problem with the SILValue API since even though one can
get an instruction or an insertion point for a value, to find the appropriate
next instruction one needs to pierce through the API and see if one has a
SILArgument or SILInstruction breaking the whole point of abstraction. The
specific problem here is that a SILArgument's "next instruction" is the first
element of the block (that is ValueBase::getDefiningInsertionPoint()) and
SILInstruction's "next instruction" is
std::next(ValueBase::getDefiningInsertionPoint()). This new
API (ValueBase::getNextInstruction()) handles this case for the compiler writer
and eliminates unnecessary code contortions.

I also did a little cleanup where I moved a doxygen comment from a near by a
const_casting trampoline method to the method that the trampoline called (see
getDefiningInsertionPoint()).

gottesmm commented:
@atrick first batch is just helpers.

gottesmm commented:
Going to be adding more on top though. I'll tell you when it is ready.

gottesmm committed:
[ownership] Rename OwnershipFixupContext::replaceAllUsesAndErase{FixingOwnership,}.

The class name is already OwnershipFixupContext... why do we need to include
FixingOwnership in its helpers... its redundant.

gottesmm committed:
[sil] Move swift::replaceAllUsesAndErase from Analysis/SimplifyInstruction -> Utils/InstOptUtils

This is a low level API already being used in multiple places besides
InstSimplify (e.x.: Utils/OwnershipOptUtils), so it makes sense to move it into
InstOptUtil.

gottesmm committed:
[sil-ownershipoptutils] Move some helper methods to their own file section from replaceAllUsesAndErase's section before adding replaceSingleUse.

Both APIs may use these APIs so it makes sense to move them to their own
section.

atrick commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -531,6 +533,22 @@ class InteriorPointerOperandKind {
+  /// Given a \p value that is a result of an instruction with an interior
+  /// pointer operand, return the interior pointer operand kind that would be
+  /// appropriate for that operand.
+  static Optional<InteriorPointerOperandKind> inferFromResult(SILValue value) {
```
The `Optional` is just an extra layer of abstraction and overhead that isn't more helpful than discriminating Invalid cases in the enum.

gottesmm replied:
It is useful in the sense that it documents in the code that one needs to handle failure there and prevents you from not handling it.

gottesmm replied:
Ok. I will change this.

atrick commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -546,14 +564,40 @@ struct InteriorPointerOperand {
+  /// If \p val is a result of an instruction that is an interior pointer,
+  /// return an interor pointer operand based off of the base value operand of
+  /// the instruction.
+  static Optional<InteriorPointerOperand>
```
I dislike passing around and unwrapping optionals when the abstraction can have an invalid representation.

gottesmm replied:
Ok. I see your point. It would be great if c++ would let us do guard let like in swift = ). That would make this not an issue.

atrick commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -569,6 +613,11 @@ struct InteriorPointerOperand {
   }
 
+  /// Return the base BorrowedValue of the incoming value's operand.
+  Optional<BorrowedValue> getSingleBaseValue() const {
```
It's very efficient, and works very well in practice for `SILValue()` to have an invalid default state. Why introduce a different programming model for `BorrowedValue` involving Optional when you want them to be used the same way?

gottesmm replied:
I think what you are saying here makes sense. I'll change that.

atrick reviewed (commented):
Commentary on my dislike of `Optional<T>` when it isn't needed.

gottesmm committed:
[ownership] Change canFixUpOwnershipForRAUW so that 'oldValue' is a SILValue not a SingleValueInstruction

The only operational change here is that I needed to be able to grab the module
from the SILValue so I could see if we were in Raw SIL or not. I realized the
only case where we could not get the module is from SILUndef and at this point
in the code we know we are going to bail already for SILUndef. This is because
we already know our new value doesn't have OwnershipKind::None and we don't
replace OwnershipKind::None things with non-OwnershipKind::None things since I
haven't implemented support for that corner case yet (but will with time).

Once I realized the previous paragraph, I was able to add support without issue.

gottesmm committed:
[sil-inst-opt] Add a new utility swift::replaceSingleUse.

Given an Operand *op, this API executes op->set(newSILValue) except that:

1. If the user of op is an end scope, this API no-opts. This API is only used in
   contexts where we are rewriting uses and are not interesting in end scope
   instructions since we are moving uses from one scope to another scope.

2. If the user of op is not an end scope, but is a lifetime ending use of
   op->get(), we insert a destroy_value|end_borrow as appropriate on op->get()
   to ensure op->get()'s lifetime is still ended. We assume that if
   op->getUser() is lifetime ending, that our caller has ensured that we can end
   newValue's lifetime.

gottesmm committed:
[sil-combine] Implement a new class SingleBlockOwnedForwardingInstFolder to help eliminate chains of owned forwarding insts and use it to update upcast and unchecked_ref_cast for ownership.

b644c80f90fb7099ec956bb44065b50e432c5146 caused all owned forwarding
instructions to be sunk to their uses in the exact cases where we could
eliminate the parent forwarding inst (namely that the value we want to fold has
no non-debug, non-consuming users). So I was able to implement this just by
implementing a single basic block algorithm that works via a planner struct
using said canonicalization. One initializes the planner struct with the
instruction that is going to either be eliminated or have its forwarding operand
set. Then one adds each of the individual chains that lead to the use that we
wish to fold, each time checking that we can eliminate the instruction.

Once the user has added all of the intermediate forwarding instructions, by
construction (see paragraph above), we know we can optimize. So we eliminate all
intermediate values and then depending on whether the user called the set value
or replace value method we either set front's operand to be the passed in value
or we RAUW/erase front with that value. It is important to note that before we
do one of those two operations, front's operand is undef, so we need to perform
one of these two operations.

gottesmm committed:
[ownership] Give better names to methods on OwnershipLifetimeExtender.

Specifically,

copyAndExtendForLifetimeEndingRAUW -> createPlusOneCopy(value, oldConsumingUse)
copyAndExtendForNonLifetimeEndingRAUW -> createPlusZeroCopy(value, nonLifetimeEndingUses)
copyBorrowAndExtendForRAUW -> createPlusZeroBorrow(value, nonLifetimeEndingUses)

Beyond being shorter, the PlusOne vs PlusZero mnemonic is telling the caller
whether or not the caller is responsible for cleaning up the returned
copy_value. In the PlusZero cases, we assume that all uses are non-lifetime
ending and we need to insert all destroy_value at the liveness points of the
value.

In the PlusOne case in contrast, we are supposed to make a copy of value and the
complete oldConsumingUse's block into a joint post-dominance set of blocks. Then
we insert destroy_values for our new copy into the completion blocks that we
found and leave cleaning up value along paths through oldConsumingUse to our
user. In this sense the value is a PlusOne value that our user must clean up for
us (sorta makes me think about ManagedValue).

gottesmm committed:
[sil-combine] Enable two tests that already pass.

gottesmm committed:
[sil-combine] Enable builtin visitors.

gottesmm committed:
[sil-combine] Update unchecked_addr_cast simplifications for ownership.

gottesmm committed:
[sil-combine] Canonicalize owned forwarding insts without non-debug non-consuming uses by sinking to their uses.

There are a bunch of optimizations in SILCombine where we try to fold an
ownership forwarding instruction A into another ownership forwarding instruction
B without deleting A. Consider the upcasts in the example below:

```
  %0 = upcast %x : $X->Y
  %1 = upcast %0 : $Y->Z
```

These sorts of optimizations fold the first instruction into the second like so:

```
  %0 = upcast %x : $X->Y
  %1 = upcast %x : $X->Z
```

This creates a problem when we are dealing with owned values since we have just
introduced two consumes for %x. To work around this, we have two options:

1. Introduce extra copies.

2. We recognize the situations where we can guarantee that we can delete the
   first upcast.

The first choice I believe is not a choice since breaking a forwarding chain of
ownership in favor of extra copies is a less canonical form. That leaves us with
the second form. What are the necessary/sufficient conditions for deleting the
first upcast. Simply it is that the upcast cannot have any non-debug,
non-consuming uses! In such a case, we know that along all paths through the
program the value has exactly one non-debug use, one of its consuming uses. If
when optimizing upcasts we could recognize that pattern, duplicate the inst
along paths not through our 2nd upcast and thus delete the original upcast
fixing the ownership error!

While this is all nice and good there is a problem with this: it doesn't
scale. As I was writing a few optimizations like this I began to note that I had
to write different versions of this same helper for many of the visitors (they
generally varied by how many forwarding instructions they looked through).

As I pondered the above, I chatted a bit with @atrick and during our
conversation, we both realized that it is much easier to solve this problem in
one block and that the condition above would allow us to sink these instructiosn
into the same block and thus if we could check for this condition and
canonicialize the IR to sink these instructions before we visiting, we could use
a single helper to handle all of these cases.

gottesmm committed:
[sil-combine] Update ownership for bridge_object_to_ref.

This just uses the infrastructure in place from fixing
upcast/unchecked_ref_cast.

gottesmm committed:
[sil-combine] Enable unchecked_trivial_bitcast for ownership.

Instruction just deals with trivial values.

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
[ownership] Rename OwnershipForwardingInst -> OwnershipForwardingMixin and eliminate support for isa<> and dyn_cast<>.

Instead, I just added some static helper methods that perform the same
operations without needing to deal with generics/etc on OwnershipForwardingMixin
itself. The reason why I did this is that this Mixin is not part of the SILNode
heirarchy so we shouldn't use utilities tied to the SILNode hierarchy.

gottesmm committed:
[ome] Explicitly set to OwnershipKind::None the ownership forwarded by all instructions that inherit from OwnershipForwardingMixin.

I thought it would be enough to change SILValue::getOwnershipKind() to return
None when we are not in OSSA, but I found that is not sufficient. In some cases,
users will call getOwnershipKind() directly on these classes and will get the
stored value (which has not been changed) instead of the correct ownership,
OwnershipKind::None.

I ran into this when one of the RAUW utilities when done a path that wanted
Owned ownership when we were in a non-ossa function and then (happily I may add)
hit an assert.

gottesmm committed:
[ownership] Ban ownership constraints with preferredKind OwnershipKind::Unowned.

It used to be that we had certain instructions that could only have operands
with OwnershipKind::{None,Unowned}. This changed with the addition of
OperandOwnership which caused us to begin accepting owned and guaranteed in all
positions where we accept unowned values with the proviso that certain
forwarding instructions are now allowed to force their result to be unowned even
though their argument is owned or guaranteed.

This setup massively simplifies RAUWing values in OSSA by eliminating the need
to insert conversion unchecked_ownership_conversion instructions from
owned->unowned when replacing an unowned value with an owned or guaranteed
value. Now, these uses can just take the owned or guaranteed value directly and
we wish to maintain that property. Thus we are banning it in SIL by writing a
verifier check.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
Things should all compile now. Not sure if all tests will pass, but things should compile.

gottesmm commented:
@swift-ci smoke test

atrick commented:
@gottesmm thanks! I completely understand the desire to use `Optional<T>`, I just think our C++ code base has a different set of tradeoffs. Namely

- The Optional template significantly decreases readability both in the declaration and at all the use sites

- The Optional template is less efficient for an already discriminated type. We don't want it on the critical path.

- In practice, the bugs we see from forgetting to check for valid values are just as easy to debug as the bugs we would get from force-unwrapping. More importantly, we should comment APIs clearly if they allow invalid values as arguments, and otherwise just assume they don't! I've seen this argument rehashed so many times. It's the same as the pointer vs. reference passing argument. Passing pointers clearly wins out in practice in terms of how much easier the code is to work with.

- build time matters and C++ template crap should only be used when it significantly improves readability/expressibility, not just because you *can* use it to be cool (same goes for closure-based generic algorithms vs simple range-based for loops to some extent)

atrick commented:
@gottesmm I left out the most important thing... I find the code much easier to navigate in lldb without extraneous template abstractions!

gottesmm committed:
[ownership] Fix duplicate getOperandRef def given upstream changes that I rebased upon.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci benchmark

gottesmm commented:
@swift-ci test windows platform

gottesmm committed:
[ownership] Fix a small thinko I made when changing InteriorPointerOperandKind to use an Invalid case instead of an Optional.

gottesmm commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -520,7 +520,6 @@ class InteriorPointerOperandKind {
   InteriorPointerOperandKind(Kind newValue) : value(newValue) {}
 
   operator Kind() const {
-    assert(isValid());
```
The reason why is that this operator is used to check if the underlying kind is invalid or not when performing a bool check.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci benchmark

gottesmm commented:
@swift-ci test source compatibility

atrick commented on lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:
```diff
@@ -95,16 +96,120 @@ SILCombiner::visitRefToRawPointerInst(RefToRawPointerInst *rrpi) {
+    // optimize only if our original value doesn't have any non-debug uses, we
+    // know that our value can only have a single non-debug use, the consuming
+    // user. So if we are not in that situation, bail.
+    if (!hasOneNonDebugUse(next))
```
Is there an assumption that `rest.back()` is the only non-debug use of `next`? If so, can you specify that and add an assert?
...
I see there is an assert later during cleanup. Still, to be able to understand how this API works, I think that invariant needs to be specified in the `add` method

gottesmm replied:
SGTM. I'll add that in my follow on.

atrick commented on lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:
```diff
@@ -95,16 +96,120 @@ SILCombiner::visitRefToRawPointerInst(RefToRawPointerInst *rrpi) {
+  /// This means that after this runs front's forwarding operand is now
+  /// SILUndef.
+  void cleanupRest() & {
+    // We process backwards -> forwards. This cleans up everything but the front
```
> We process backwards -> forwards.

I can't parse this sentence. I think you mean we process top-down in the def-use chain

gottesmm replied:
yes

gottesmm replied:
I'll fix the comment.

atrick commented on lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:
```diff
@@ -95,16 +96,120 @@ SILCombiner::visitRefToRawPointerInst(RefToRawPointerInst *rrpi) {
+    // forwarding operand set to Undef, we haven't touched its result. So it is
+    // safe to RAUW.
+    cleanupRest();
+    SC.replaceValueUsesWith(front, newValue);
```
Question: should we also be erasing `front` here since it returns `nullptr`?

gottesmm replied:
I did this on purpose since I wanted to signal to the caller/SILCombine not to do anything further. I took care of everything.

atrick replied:
@gottesmm my point is that if you are taking care of everything (hence returning nullptr) shouldn't you be erasing the dead `front` instruction?

atrick commented on lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:
```diff
@@ -282,35 +387,92 @@ SILCombiner::visitUncheckedAddrCastInst(UncheckedAddrCastInst *UADCI) {
+    if (folder.add(ui)) {
+      auto *newValue =
+          Builder.createUncheckedRefCast(urci->getLoc(), uiOp, urci->getType());
+      return std::move(folder).optimizeWithReplacement(newValue);
```
Why are you creating a new instruction here and using `optimizeWithReplacement`?

gottesmm replied:
I am matching what the code did previously.

atrick commented on lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:
```diff
@@ -282,35 +387,92 @@ SILCombiner::visitUncheckedAddrCastInst(UncheckedAddrCastInst *UADCI) {
+    if (folder.add(otherURCI)) {
+      auto *newValue = Builder.createUncheckedRefCast(
+          urci->getLoc(), otherURCIOp, urci->getType());
+      return std::move(folder).optimizeWithReplacement(newValue);
```
I'm confused about why you're creating a new instruction here.

gottesmm replied:
I am matching what the code did previously.

atrick commented on lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:
```diff
@@ -282,35 +387,92 @@ SILCombiner::visitUncheckedAddrCastInst(UncheckedAddrCastInst *UADCI) {
+      SingleBlockOwnedForwardingInstFolder folder(*this, urci);
+      if (folder.add(oer) && folder.add(ier)) {
+        auto *newValue = Builder.createUncheckedRefCast(
+            urci->getLoc(), ier->getOperand(), urci->getType());
```
Still not sure why you create a new instruction

gottesmm replied:
Also matching what the code did previously. It wasn't clear to me, why this needed to be done, so I just matched it.

atrick reviewed (approved):
There are several follow-up comments for the next round of SILCombine. I mainly got hung up on the new instruction folder where I'm either missing something or it could be clarified.

gottesmm commented:
Andy and I spoke about this, I am going to fix his comments in another PR. This PR is too big and is unwieldy

gottesmm commented on include/swift/SIL/SILValue.h:
```diff
@@ -468,11 +468,46 @@ class ValueBase : public SILNode, public SILAllocated<ValueBase> {
+  /// Operationally this means that:
+  ///
+  /// * For SILArguments, this returns the first instruction in the block. This
+  ///   is the main divergence from getDefiningInsertionPoint (see discussion
```
For the comment fix: #35431

gottesmm commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -734,9 +734,26 @@ bool OwnershipFixupContext::canFixUpOwnershipForRAUW(
 
+  // First check if oldValue is SILUndef. If it is, then we know that:
+  //
+  // 1. SILUndef (and thus oldValue) must have OwnershipKind::None.
```
@atrick I looked at this. It looks like there is some sort of sentinel value that is used in certain places so I am going to have to figure out a way to bridge that. I do think it is a silly footman to leave in the code base.

gottesmm commented on lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:
```diff
@@ -127,11 +125,20 @@ visitPointerToAddressInst(PointerToAddressInst *PTAI) {
-      return Builder.createUncheckedAddrCast(PTAI->getLoc(), ATPI->getOperand(),
-                                             PTAI->getType());
+    // We can not perform this optimization with ownership until we are able to
+    // handle issues around interior pointers and expanding borrow scopes.
```
Yes. I am preparing a PR that handles this by adding an interior pointer fixing up helper for this.

gottesmm commented on lib/SILOptimizer/SILCombiner/SILCombine.cpp:
```diff
@@ -145,6 +152,63 @@ class SILCombineCanonicalize final : CanonicalizeInstruction {
+    svi->moveBefore(consumingUser);
+    MadeChange = true;
+
+    // NOTE: We return nullptr here so that our caller doesn't delete the
```
Oops... I'll fix.

gottesmm commented on lib/SILOptimizer/SILCombiner/SILCombiner.h:
```diff
@@ -347,9 +353,26 @@ class SILCombiner :
-private:
+  /// Let \p user and \p value be two forwarding single value instructions with
+  /// the property that \p user, through potentially a chain of forwarding
+  /// instructions.
```
I think I forgot to delete this and started below. I will eliminate this part.
